src/merkle/partial_mt/mod.rs

```diff
@@ -0,0 +1,306 @@
+// ================================================================================================
+
+/// A partial Merkle tree with NodeIndex keys and 4-element RpoDigest leaf values. Partial Merkle
+/// Tree allows to create Merkle Tree by providing Merkle paths of different lengths.
+///
+/// The root of the tree is recomputed on each new leaf update.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PartialMerkleTree {
+    max_depth: u8,
+    nodes: BTreeMap<NodeIndex, RpoDigest>,
+    leaves: BTreeSet<NodeIndex>,
+}
+
+impl Default for PartialMerkleTree {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl PartialMerkleTree {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    /// Minimum supported depth.
+    pub const MIN_DEPTH: u8 = 1;
+
+    /// Maximum supported depth.
+    pub const MAX_DEPTH: u8 = 64;
+
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a new emply [PartialMerkleTree].
+    pub fn new() -> Self {
+        PartialMerkleTree {
+            max_depth: 0,
+            nodes: BTreeMap::new(),
+            leaves: BTreeSet::new(),
+        }
+    }
+
+    /// Appends the provided paths iterator into the set.
+    ///
+    /// Analogous to [Self::add_path].
+    pub fn with_paths<I>(paths: I) -> Result<Self, MerkleError>
+    where
+        I: IntoIterator<Item = (u64, RpoDigest, MerklePath)>,
+    {
+        // create an empty tree
+        let tree = PartialMerkleTree::new();
+
+        paths.into_iter().try_fold(tree, |mut tree, (index, value, path)| {
+            tree.add_path(index, value, path)?;
+            Ok(tree)
+        })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the root of this Merkle tree.
+    pub fn root(&self) -> RpoDigest {
+        self.nodes.get(&ROOT_INDEX).cloned().unwrap_or(EMPTY_DIGEST)
+    }
+
+    /// Returns the depth of this Merkle tree.
+    pub fn max_depth(&self) -> u8 {
+        self.max_depth
+    }
+
+    /// Returns a node at the specified NodeIndex.
+    ///
+    /// # Errors
+    /// Returns an error if the specified NodeIndex is not contained in the nodes map.
+    pub fn get_node(&self, index: NodeIndex) -> Result<RpoDigest, MerkleError> {
+        self.nodes.get(&index).ok_or(MerkleError::NodeNotInSet(index)).map(|hash| *hash)
+    }
+
+    /// Returns true if provided index contains in the leaves set, false otherwise.
+    pub fn is_leaf(&self, index: NodeIndex) -> bool {
+        self.leaves.contains(&index)
+    }
+
+    pub fn get_leaf_depth(&self, index: u64) -> u8 {
+        let trunc_value = index.min(2_u64.pow(self.max_depth() as u32));
+        let mut node_index =
+            NodeIndex::new(self.max_depth(), trunc_value).expect("Truncated value is not valid.");
```

Comment:
I'd do it differently here: first we can instantiate node index using `depth = 64`, and then we can use `NodeIndex::move_up_to()` to move to `max_depth`.

Reply:
With this approach we will have difficulties with calculating value of the node index. Since `move_up_to` function changes not only depth, but also value, dividing it by `2^(max_depth - actual_depth)`, we need to calculate appropriate value for the node index with max depth. So the value of the `NodeIndex` with `depth = 64` will be 
`index * 2^(max_depth - actual_depth)`. So this approach will require calculating the value twice. That's not a big deal, but is it worth it?

Reply:
I'm not sure I follow. I was thinking we could do something like:
```Rust
let mut node_index = NodeIndex::new(64, index).expect("");
node_index.move_up_to(self.max_depth());
```
Or is this not sufficient for some reason?

Reply:
Ah - I see! I wonder if we should re-think this method somewhat. Maybe the `index` it takes should already be `NodeIndex` which specifies both value and depth. Otherwise, it seems like we may be getting unintuitive results from it.